source/Missions/Services/Agents/Handlers/WeaponPickupHandler.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Autofac;
+using Common.Messaging;
+using Common.Network;
+using GameInterface.Serialization;
+using HarmonyLib;
+using Missions.Services.Agents.Messages;
+using Missions.Services.Network;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using TaleWorlds.Core;
+using TaleWorlds.MountAndBlade;
+
+namespace Missions.Services.Agents.Handlers
+{
+    /// <summary>
+    /// Handler for weapon pickups within a battle
+    /// </summary>
+    public interface IWeaponPickupHandler : IHandler, IDisposable
+    {
+
+    }
+    /// <inheritdoc/>
+    public class WeaponPickupHandler : IWeaponPickupHandler
+    {
+        readonly INetworkAgentRegistry networkAgentRegistry;
+        readonly INetworkMessageBroker networkMessageBroker;
+        public WeaponPickupHandler(INetworkAgentRegistry networkAgentRegistry, INetworkMessageBroker networkMessageBroker)
+        {
+            this.networkAgentRegistry = networkAgentRegistry;
+            this.networkMessageBroker = networkMessageBroker;
+
+            networkMessageBroker.Subscribe<WeaponPickedup>(WeaponPickupSend);
+            networkMessageBroker.Subscribe<NetworkWeaponPickedup>(WeaponPickupReceive);
+
+        }
+        ~WeaponPickupHandler()
+        {
+            Dispose();
+        }
+
+        public void Dispose()
+        {
+            networkMessageBroker.Unsubscribe<WeaponPickedup>(WeaponPickupSend);
+            networkMessageBroker.Unsubscribe<NetworkWeaponPickedup>(WeaponPickupReceive);
+        }
+
+        private void WeaponPickupSend(MessagePayload<WeaponPickedup> obj)
+        {
+            if (!networkAgentRegistry.IsControlled(obj.What.Agent)) return;
```

Comment:
Instead of using `!` use `== false` instead. I'm going to add it to our coding standard whenever I get to that
```suggestion
            if (networkAgentRegistry.IsControlled(obj.What.Agent) == false) return;
```